sdklab/regressions/regression_github_990.py

```diff
@@ -0,0 +1,284 @@
+    def new_reconnect(*args, **kwargs):
+        """
+        `reconnect` needs to raise an exception when we need it to in order to simulate the exception
+        that lead to the discovery of this bug.
+        """
+        nonlocal raise_on_next_reconnect
+        if raise_on_next_reconnect:
+            log_func(
+                "----------- RECONNECT CALLED. raising {}".format(
+                    type(exception_to_raise_on_reconnect)
+                )
+            )
+            raise_on_next_reconnect = False
+            raise exception_to_raise_on_reconnect
+        return old_reconnect(*args, **kwargs)
+
+    paho._sock_send = new_sock_send
+    paho._sock_recv = new_sock_recv
+    paho._sock_close = new_sock_close
+    paho.publish = new_publish
+    paho.reconnect = new_reconnect
+
+
+async def run_test(
+    connection_retry, auto_connect, send_should_succeed, exception_to_raise_on_reconnect
+):
+    try:
+        print("*" * 80)
+        print()
+        print(
+            "Running test with connection_retry={}, auto_connect={}, and exception_to_raies_on_reconnect={}".format(
+                connection_retry, auto_connect, type(exception_to_raise_on_reconnect)
+            )
+        )
+        print()
+        print("*" * 80)
+
+        # Create instance of the device client using the connection string
+        device_client = IoTHubDeviceClient.create_from_connection_string(
+            test_env.DEVICE_CONNECTION_STRING,
+            keep_alive=10,
+            connection_retry=connection_retry,
+            auto_connect=auto_connect,
+        )
+        logging_hook.hook_device_client(device_client)
+
+        hack_paho_to_disconnect_after_publish(device_client, exception_to_raise_on_reconnect)
+
+        # Connect the device client.
+        await device_client.connect()
+
+        try:
+            print("Sending message...")
+            await device_client.send_message("This is a message that is being sent")
+            print("Message successfully sent!")
+            assert send_should_succeed
+        except Exception as e:
+            if send_should_succeed:
+                raise
+            else:
+                print("send_message failed as expected.")
+                print("raised: {}".format(str(e) or type(e)))
+
+        print("Shutting down")
+        await device_client.shutdown()
+
+    except Exception:
+        print("FAILED " * 10)
+        print(
+            "FAILED with connection_retry={}, auto_connect={}, and exception_to_raies_on_reconnect={}".format(
+                connection_retry, auto_connect, type(exception_to_raise_on_reconnect)
+            )
+        )
+        raise
+
+
+async def main():
+    workaround_github_990()
+
+    # test with retryable errors.
+    tls_auth_error = ssl.SSLError()
+    tls_auth_error.strerror = "CERTIFICATE_VERIFY_FAILED"
+
+    # If `connection_retry` is `True`, the client should re-connect and
+    # `send_message` should succeed
+    await run_test(
```

Comment:
Doesn't running a sequence of these tests encounter the same problem you wished to solve by going outside of pytest? sys exit will only happen at the very end of all tests, not once per test.

Reply:
That's true, but PR 1023 was for a shutdown-specific bug, and I don't want to fall into the trap of "we found one shutdown bug that could only repro if you shutdown and then exited, so we should make all of our test paths shutdown and then exit."